### PR TITLE
extending callbacks to gan

### PR DESCRIPTION
**Pull request by hrishikeshSuresh**
Referring issue #2013

### CODE REVIEW

**Comment by hrishikeshSuresh** [comment]
Can I add callbacks to some other methods with this PR itself?

**Comment by zoq** [comment]
Hello @hrishikeshSuresh thanks for the contribution, I would open a new PR, makes the review a lot easier.

File: src/mlpack/tests/gan_test.cpp [hunk 1]
```diff
@@ -215,7 +215,8 @@ BOOST_AUTO_TEST_CASE(GANMNISTTest)
       discriminatorPreTrain, multiplier);
 
   Log::Info << "Training..." << std::endl;
-  double objVal = gan.Train(optimizer);
+  std::stringstream stream;
+  double objVal = gan.Train(optimizer, ens::PrintLoss(stream));
```
Comment (zoq) [inline src/mlpack/tests/gan_test.cpp]: We should check that the callback works, e.g. by checking if the length of the `stream` parameter isn't empty.

File: src/mlpack/methods/ann/gan/gan.hpp [hunk 2]
```diff
@@ -104,8 +104,9 @@ class GAN
    *
    * @return The final objective of the trained model (NaN or Inf on error).
    */
-  template<typename OptimizerType>
-  double Train(OptimizerType& Optimizer);
+  template<typename OptimizerType, typename... CallbackTypes>
+  double Train(OptimizerType& Optimizer, 
+               CallbackTypes&&... callbacks);
```
Comment (zoq) [inline src/mlpack/methods/ann/gan/gan.hpp]: Can you add a parameter description for the `callbacks` parameter?

File: src/mlpack/tests/gan_test.cpp (hunk 1 above)
Comment (hrishikeshSuresh) [reply src/mlpack/tests/gan_test.cpp]: done.

File: src/mlpack/methods/ann/gan/gan.hpp (hunk 2 above)
Comment (hrishikeshSuresh) [reply src/mlpack/methods/ann/gan/gan.hpp]: done.

**Comment by zoq** [comment]
Looks like the appveyor timed out, let me restart the build.

File: src/mlpack/methods/ann/gan/gan.hpp [hunk 3]
```diff
@@ -101,11 +101,13 @@ class GAN
 
   /**
    * Train function.
-   *
+   * 
+   * @tparam CallbackTypes Types of Callback functions.
```
Comment (zoq) [inline src/mlpack/methods/ann/gan/gan.hpp]: I know you haven"t add the parameter but can you add a description for the `OptimizerType` as well? You can copy the description from:

 https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/logistic_regression/logistic_regression.hpp#L125

Comment (hrishikeshSuresh) [reply src/mlpack/methods/ann/gan/gan.hpp]: done.

**Comment by zoq** [comment]
Also, up to you if you like to fix the merge conflicts, but we can also fix them in the process.

**Comment by hrishikeshSuresh** [comment]
I tried solving the merge conflict & then the build failed. I'll revert to the previous commit and leave it as it is or should I fix this build failure? @zoq

File: src/mlpack/methods/ann/gan/gan_impl.hpp [hunk 4]
```diff
@@ -217,14 +217,14 @@ template<
...
+template<typename OptimizerType, typename... CallbackTypes>
+double GAN<Model, InitializationRuleType, Noise,
+           PolicyType>::Train(OptimizerType& Optimizer,
+           CallbackTypes&&... callbacks)
 {
-  ResetData(std::move(trainData));
-
-  return Optimizer.Optimize(*this, parameter);
+  if (!reset)
+    Reset();
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/gan/gan_impl.hpp]: Why this change?

Comment (hrishikeshSuresh) [reply src/mlpack/methods/ann/gan/gan_impl.hpp]: Minor mistake on my side. I fixed it now. But it looks like the travis build timed out again.

File: src/mlpack/methods/ann/gan/gan_impl.hpp [hunk 5]
```diff
@@ -217,14 +217,14 @@ template<
...
   typename PolicyType
 >
-template<typename OptimizerType>
-double GAN<Model, InitializationRuleType, Noise, PolicyType>::Train(
-    arma::mat trainData,
-    OptimizerType& Optimizer)
+template<typename OptimizerType, typename... CallbackTypes>
+double GAN<Model, InitializationRuleType, Noise,
+           PolicyType>::Train(arma::mat trainData, OptimizerType& Optimizer,
+           CallbackTypes&&... callbacks)
```
Comment (walragatver) [inline src/mlpack/methods/ann/gan/gan_impl.hpp]: Small style issue over here :)
https://github.com/mlpack/mlpack/wiki/DesignGuidelines

Comment (hrishikeshSuresh) [reply src/mlpack/methods/ann/gan/gan_impl.hpp]: done.
